### PR TITLE
Ingest metric data with COPY to get better performance

### PR DESCRIPTION
**Pull request by niksajakovljevic**
We try COPY into metric table. If there is a conflict situation
(eg. duplicate key violation) we catch it on client and retry
by COPYing to temporary table and doing ON CONFLICT insert from there.

### CODE REVIEW

**Comment by niksajakovljevic** [comment]
Due to recent changes in the extension I needed to temporarily tweak `EXTENSION` to be able to run e2e tests. I believe long term we want to trigger e2e tests on the new branch from the extension side.

File: pkg/pgmodel/ingestor/copier.go [hunk 1]
```diff
@@ -310,17 +339,31 @@ func debugInsert() {
...
 	totalSamples := 0
 	totalExemplars := 0
+	var sampleRows [][]interface{}
+	var exemplarRows [][]interface{}
+	insertStart := time.Now()
 	lowestEpoch := pgmodel.SeriesEpoch(math.MaxInt64)
 	lowestMinTime := int64(math.MaxInt64)
+	tx, err := conn.BeginTx(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to start TX for inserting metrics: %v", err), lowestMinTime
```
Comment (Harkishen-Singh) [inline pkg/pgmodel/ingestor/copier.go:357]:
```suggestion
		return fmt.Errorf("failed to start transaction for inserting metrics: %w", err), lowestMinTime
```

**Review by Harkishen-Singh — CHANGES_REQUESTED** [review]
The implementation looks good. We should have some end-to-end test, that can stimulate normal COPY ingest and with conflict.

**Comment by niksajakovljevic** [comment]
@Harkishen-Singh There are existing e2e tests for testing various duplicate scenarios (check `create_test.go`).

**Review by Harkishen-Singh — APPROVED** [review]

File: pkg/pgmodel/ingestor/copier.go [hunk 2]
```diff
@@ -403,63 +425,96 @@ func insertSeries(ctx context.Context, conn pgxconn.PgxConn, reqs ...copyRequest
...
-			// We cannot send 2-D [][]TEXT to postgres via the pgx.encoder. For this and easier querying reasons, we create a
-			// new type in postgres by the name prom_api.label_value_array and use that type as array (which forms a 2D array of TEXT)
-			// which is then used to push using the unnest method apprach.
-			labelValues := pgmodel.GetCustomType(pgmodel.LabelValueArray)
-			if err := labelValues.Set(exemplarLbls); err != nil {
-				return fmt.Errorf("setting prom_api.label_value_array[] value: %w", err), lowestMinTime
-			}
 			numRowsPerInsert = append(numRowsPerInsert, numExemplars)
-			batch.Queue("SELECT _prom_catalog.insert_exemplar_row($1::NAME, $2::TIMESTAMPTZ[], $3::BIGINT[], $4::prom_api.label_value_array[], $5::DOUBLE PRECISION[])", req.info.TableName, timeExemplars, seriesIdExemplars, labelValues, valExemplars)
+			table := pgx.Identifier{schema.PromDataExemplar, req.info.TableName}
```
[the remark this answers is not in the document]
Comment (niksajakovljevic) [reply pkg/pgmodel/ingestor/copier.go:451]: Yep, will try to improve that part a bit

File: pkg/pgmodel/ingestor/copier.go [hunk 3]
```diff
@@ -403,63 +423,96 @@ func insertSeries(ctx context.Context, conn pgxconn.PgxConn, reqs ...copyRequest
 		totalExemplars += numExemplars
 		if hasSamples {
 			numRowsPerInsert = append(numRowsPerInsert, numSamples)
-			batch.Queue("SELECT _prom_catalog.insert_metric_row($1, $2::TIMESTAMPTZ[], $3::DOUBLE PRECISION[], $4::BIGINT[])", req.info.TableName, timeSamples, valSamples, seriesIdSamples)
+			table := pgx.Identifier{req.info.TableSchema, req.info.TableName}
+			if onConflict {
+				// we append table prefix to make sure that temp table name is unique
+				table, err = createTempIngestTable(ctx, tx, req.info.TableName, req.info.TableSchema, fmt.Sprintf("s%d_", r))
```
[the remark this answers is not in the document]
Comment (niksajakovljevic) [reply pkg/pgmodel/ingestor/copier.go:429]: I don't think that's true. Temporary tables are bound to a session/connection. That is by design.

[the remark this answers is not in the document]
Comment (niksajakovljevic) [reply pkg/pgmodel/ingestor/copier.go:429]: Related discussion https://github.com/timescale/promscale_extension/pull/401#discussion_r916747127

File: pkg/pgmodel/ingestor/copier.go [hunk 4]
```diff
@@ -403,63 +423,96 @@ func insertSeries(ctx context.Context, conn pgxconn.PgxConn, reqs ...copyRequest
...
+				if err != nil {
+					return err, lowestMinTime
+				}
+			}
+			inserted, err := tx.CopyFrom(ctx, table, schema.PromExemplarColumns[:], pgx.CopyFromRows(exemplarRows))
+			if err != nil {
+				return err, lowestMinTime
+			}
+			if onConflict {
+				res, err := tx.Exec(ctx, fmt.Sprintf(sqlInsertIntoFrom, schema.PromDataExemplar, pgx.Identifier{req.info.TableName}.Sanitize(),
```
Comment (antekresic) [inline pkg/pgmodel/ingestor/copier.go:464]: Minor nit: if the argument list is multiline, lets put all of them on new lines. Easier to read.

**Review by antekresic — APPROVED** [review]
Minor nit, looks good otherwise :tada:

**Pull request merged**